src/librustc_trans/trans/context.rs

```diff
@@ -977,11 +942,62 @@ fn declare_intrinsic(ccx: &CrateContext, key: &str) -> Option<ValueRef> {
                 ccx.intrinsics().borrow_mut().insert($name, f.clone());
                 return Some(f);
             }
+        );
+        ($name:expr, $cname:ident ($($arg:expr),*) -> $ret:expr) => (
+            if !is_like_pnacl {
+                // The `if key == $name` is already in ifn!
+                ifn!($name, fn($($arg),*) -> $ret);
+            } else if key == $name {
+                let f = declare::declare_cfn(ccx, stringify!($cname),
+                                             Type::func(&[$($arg),*], &$ret),
+                                             ccx.tcx().mk_nil());
+                ccx.intrinsics().borrow_mut().insert($name, f.clone());
+                return Some(f);
+            }
         )
     }
 
     compatible_ifn!("llvm.assume", noop(llvmcompat_assume(i1) -> void), 6);
 
+    // These must be redirected to `libm` here so PNaCl's linker will link in
+    // these functions from `libm`.
+    compatible_ifn!("llvm.copysign.f32", copysignf(t_f32, t_f32) -> t_f32);
```

Comment:
> In pnacl what does it actually mean to "link against" libm?

It means adding `-lm` to the linker flags.

> Is there a pre-built pile of bytecode corresponding to libm?

Yes. All "system" libraries have this treatment for PNaCl/JS.

> Are symbols recognized and redirected to JS functions?

Emscripten doesn't redirect the libm functions to JS.

> It still sounds to me like this is something for the linker to handle, not necessarily us in the frontend.

Except Rust is unique by the way Rust (fairly implicitly) references the builtins. Clang doesn't generate references to them unless one of the `__builtin_` functions are used, and *nobody* does that (evidenced by the fact that this case isn't handled at all by `pnacl-clang`) (or at least no common OSS does it).

You can test, if you'd like, with:
```cpp
#include <assert.h>

int main() {
  auto v = __builtin_powif(3.14, 5);
  /// will error
  assert(v == 0);
  return 0;
}
```

Compile with `pnacl-clang++ -o test.pexe test.cpp -O0`. I get:
```
Function llvm.powi.f32 is a disallowed LLVM intrinsic
LLVM ERROR: PNaCl ABI verification failed
```

Therefore I'd say it's really more of Rust's job to fix in this case.